### PR TITLE
perf(semantic): faster search for leading comments

### LINKED ISSUE #4114: semantic: huge performance regression when jsdoc parsing is enabled
https://github.com/oxc-project/oxc/runs/27005176800

This caused our CI to go from 5 minutes to > 10 minutes.

### PR DESCRIPTION
**Pull request by Boshen**
fixes #4114

@leaysgur This is my wild guess, I need a second pairs of eyes to make sure this is correct.

### CODE REVIEW

**Comment by rzvxa** [comment]
> ## [CodSpeed Performance Report](https://codspeed.io/oxc-project/oxc/branches/perf-jsdoc)
> ### Merging #4140 will **improve performances by ×59**
> 
> Comparing `perf-jsdoc` ([f5f64db](https://github.com/oxc-project/oxc/commit/f5f64dbcefdde4b002762cff567ee1362fbfa0c8)) with `main` ([714bf1d](https://github.com/oxc-project/oxc/commit/714bf1dd7f39eca7eef500039a617b093cfa591f))
> ### Summary
> 
> `⚡ 4` improvements `✅ 25` untouched benchmarks
> ### Benchmarks breakdown
> 	Benchmark 	`main` 	`perf-jsdoc` 	Change
> ⚡ 	`semantic[antd.js]` 	12,416.8 ms 	211.3 ms 	×59
> ⚡ 	`semantic[cal.com.tsx]` 	135.6 ms 	63.2 ms 	×2.1
> ⚡ 	`semantic[checker.ts]` 	2,601.6 ms 	135.7 ms 	×19
> ⚡ 	`semantic[pdf.mjs]` 	89 ms 	33.3 ms 	×2.7

WTF

**Comment by Boshen** [comment]
> > ## [CodSpeed Performance Report](https://codspeed.io/oxc-project/oxc/branches/perf-jsdoc)
> > ### Merging #4140 will **improve performances by ×59**
> > Comparing `perf-jsdoc` ([f5f64db](https://github.com/oxc-project/oxc/commit/f5f64dbcefdde4b002762cff567ee1362fbfa0c8)) with `main` ([714bf1d](https://github.com/oxc-project/oxc/commit/714bf1dd7f39eca7eef500039a617b093cfa591f))
> > ### Summary
> > `⚡ 4` improvements `✅ 25` untouched benchmarks
> > ### Benchmarks breakdown
> > Benchmark 	`main` 	`perf-jsdoc` 	Change
> > ⚡ 	`semantic[antd.js]` 	12,416.8 ms 	211.3 ms 	×59
> > ⚡ 	`semantic[cal.com.tsx]` 	135.6 ms 	63.2 ms 	×2.1
> > ⚡ 	`semantic[checker.ts]` 	2,601.6 ms 	135.7 ms 	×19
> > ⚡ 	`semantic[pdf.mjs]` 	89 ms 	33.3 ms 	×2.7
> 
> WTF

Just a normal day in oxc ...

File: crates/oxc_semantic/src/jsdoc/builder.rs [hunk 1]
```diff
@@ -117,9 +120,7 @@ impl<'a> JSDocBuilder<'a> {
 
         let span = kind.span();
         let mut leading_comments = vec![];
-        // May be better to set range start for perf?
-        // But once I tried, coverage tests start failing...
-        for comment in self.trivias.comments_range(..span.start) {
+        for comment in self.trivias.comments_range(self.previous_span_end..span.start) {
```
Comment (lucab) [inline crates/oxc_semantic/src/jsdoc/builder.rs:123]: I think you can delete the `leading_comments` vector as well, and avoid iterating again on it.
It should be enough to conditionally-push directly into `leading_jsdoc_comments`:
```
let comments_len = comments_range.size_hint().1;
let mut leading_jsdoc_comments = Vec::with_capacity(comments_len.unwrap_or(0));

for (_) {
[...]
  self.leading_comments_seen.insert(comment.span.start);
  if let Some(jsdoc) = self.parse_if_jsdoc_comment(_) {
    leading_jsdoc_comments.push(jsdoc);
  }
}

[...]
leading_jsdoc_comments.shrink_to_fit();
self.attached_docs.insert(span, leading_jsdoc_comments);
```

Comment (Boshen) [reply crates/oxc_semantic/src/jsdoc/builder.rs:123]: Do you wanna give this a try? It's getting late here 😅

Comment (lucab) [reply crates/oxc_semantic/src/jsdoc/builder.rs:123]: I think I have to first figure out some issues with my Graphite setup, don't block on me.
I can add this additional change in a followup PR later.

Comment (Boshen) [reply crates/oxc_semantic/src/jsdoc/builder.rs:123]: You may also push to this branch. Need to wait for a review from the previous author anyway.

**Comment by leaysgur** [comment]
> This is my wild guess, I need a second pairs of eyes to make sure this is correct.

LGTM!

https://github.com/oxc-project/oxc/pull/4140/files#diff-79d6155264aad0a2e14ac20754321839cbd6ef3735ab83c7238d4e2d613d45efL120-L121

Even though the code seems familiar, why didn't it work that time... 😅

**Pull request merged**